docs/branching-and-tagging.md

```diff
@@ -0,0 +1,218 @@
+ - this work for additional information regarding copyright ownership.
+ - The ASF licenses this file to You under the Apache License, Version 2.0
+ - (the "License"); you may not use this file except in compliance with
+ - the License.  You may obtain a copy of the License at
+ -
+ -   http://www.apache.org/licenses/LICENSE-2.0
+ -
+ - Unless required by applicable law or agreed to in writing, software
+ - distributed under the License is distributed on an "AS IS" BASIS,
+ - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ - See the License for the specific language governing permissions and
+ - limitations under the License.
+ -->
+
+# Branching and Tagging
+
+## Overview
+
+Iceberg table metadata maintains a log of snapshots which represent the changes applied to a table. Snapshots are fundamental in Iceberg as they are the basis for reader isolation and time travel queries. For controlling metadata size and storage costs, Iceberg provides snapshot lifecycle management procedures such as [`expireSnapshots`](../../../javadoc/{{% icebergVersion %}}/org/apache/iceberg/Table.html#expireSnapshots--) for removing unused snapshots and no longer neccessary data files based on table snapshot retention properties.
+
+**For more sophisticated snapshot lifecycle management, Iceberg also supports branches and tags which are named references to snapshots.** Branches are independent lineages of snapshots and point to the head of the lineage. Branches and tags have independent lifecycles with regards to snapshot retention during expireSnapshots.
+
+Branching and tagging can be used for handling GDPR requirements and retaining important historical snapshots for auditing. Branches can also be used as part of data engineering workflows, for enabling experimental branches for testing and validating new jobs. New jobs can write to the branch without impacting the main table state. After validation of the new job, the main branch can be fast forwarded. 
+
+Branches and tags have a maximum reference age property 
+which control when the reference to the snapshot itself should be expired. Branches have retention properties which define the minimum number of snapshots to retain on a branch as well as the maximum age of individual snapshots to retain on the branch. These properties are used when the expireSnapshots procedure is run. For details on the algorithm for expireSnapshots, refer to the [spec](../../../spec/#snapshot-retention-policy).
+
+## Usage 
+
+### Iceberg Java Library
+
+The Iceberg Java library can be used for performing operations on branches and tags.
+
+#### Creating branches and tags
+
+New branches and tags can be created via the Java library's ManageSnapshots API. 
+
+```java
+import org.apache.iceberg.Table;
+
+Table table = catalog.loadTable(...);
+
+/* Create a branch test-branch which is retained for 1 week, and the latest 2 snapshots on test-branch will always be retained. 
```

Comment:
nit: I think it would help to mention and include `setMaxSnapshotAgeMs` / `setMaxRefAgeMs` / ... here in the text and explain what refers to what

Reply:
Makes sense, done!